File: src/routes/examples/norns/todo/store.c
import { writable, derived } from 'svelte/store'

todos = writable []

addTodo = (text) ->
	text = text.trim()
	return unless text
	todos.update (list) ->
		[...list, { id: crypto.randomUUID(), text, done: false }]

toggleTodo = (id) ->
	todos.update (list) ->
		list.map (t) ->
			if t.id is id then { ...t, done: not t.done } else t

removeTodo = (id) ->
	todos.update (list) ->
		list.filter (t) -> t.id isnt id

clearCompleted = ->
	todos.update (list) ->
		list.filter (t) -> not t.done

remaining = derived todos, ($todos) ->
	$todos.filter((t) -> not t.done).length

total = derived todos, ($todos) -> $todos.length

export { todos, addTodo, toggleTodo, removeTodo, clearCompleted, remaining, total }

File: src/routes/examples/svelte/todo/store.ts
import { writable, derived } from 'svelte/store';

export type Todo = { id: string; text: string; done: boolean };

export const todos = writable<Todo[]>([]);

export function addTodo(text: string): void {
	text = text.trim();
	if (!text) return;
	todos.update((list) => [...list, { id: crypto.randomUUID(), text, done: false }]);
}

export function toggleTodo(id: string): void {
	todos.update((list) =>
		list.map((t) => (t.id === id ? { ...t, done: !t.done } : t))
	);
}

export function removeTodo(id: string): void {
	todos.update((list) => list.filter((t) => t.id !== id));
}

export function clearCompleted(): void {
	todos.update((list) => list.filter((t) => !t.done));
}

export const remaining = derived(todos, ($todos) => $todos.filter((t) => !t.done).length);
export const total = derived(todos, ($todos) => $todos.length);
